sidecar/src/service/blocking.rs

```diff
@@ -6,22 +6,92 @@ use super::{
     SidecarInterfaceRequest, SidecarInterfaceResponse,
 };
 use crate::dogstatsd::DogStatsDAction;
-use datadog_ipc::platform::ShmHandle;
+use datadog_ipc::platform::{Channel, ShmHandle};
 use datadog_ipc::transport::blocking::BlockingTransport;
+use std::sync::Mutex;
 use std::{
     borrow::Cow,
     io,
     time::{Duration, Instant},
 };
+use tracing::info;
 
-/// `SidecarTransport` is a type alias for the `BlockingTransport` struct from the `datadog_ipc`
-/// crate. It is used for sending `SidecarInterfaceRequest` and receiving
-/// `SidecarInterfaceResponse`.
+/// `SidecarTransport` is a wrapper around a BlockingTransport struct from the `datadog_ipc` crate
+/// that handles transparent reconnection.
+/// It is used for sending `SidecarInterfaceRequest` and receiving `SidecarInterfaceResponse`.
 ///
 /// This transport is used for communication between different parts of the sidecar service.
 /// It is a blocking transport, meaning that it will block the current thread until the operation is
 /// complete.
-pub type SidecarTransport = BlockingTransport<SidecarInterfaceResponse, SidecarInterfaceRequest>;
+pub struct SidecarTransport {
+    pub inner: Mutex<BlockingTransport<SidecarInterfaceResponse, SidecarInterfaceRequest>>,
+}
+
+impl SidecarTransport {
+    pub fn reconnect<F>(&mut self, factory: F)
+    where
+        F: FnOnce() -> Option<Box<SidecarTransport>>,
+    {
+        let mut transport = match self.inner.lock() {
+            Ok(t) => t,
+            Err(_) => return,
+        };
+        if transport.is_closed() {
+            info!("The sidecar transport is closed. Reconnecting...");
+            let new = match factory() {
+                None => return,
+                Some(n) => n.inner.into_inner(),
+            };
+            if new.is_err() {
+                return;
+            }
+            *transport = new.unwrap();
+        }
+    }
+
+    pub fn set_read_timeout(&mut self, timeout: Option<Duration>) -> io::Result<()> {
+        match self.inner.lock() {
+            Ok(mut t) => t.set_read_timeout(timeout),
+            Err(e) => Err(io::Error::new(io::ErrorKind::Other, e.to_string())),
+        }
+    }
+
+    pub fn set_write_timeout(&mut self, timeout: Option<Duration>) -> io::Result<()> {
+        match self.inner.lock() {
+            Ok(mut t) => t.set_write_timeout(timeout),
+            Err(e) => Err(io::Error::new(io::ErrorKind::Other, e.to_string())),
+        }
+    }
+
+    pub fn is_closed(&self) -> bool {
+        match self.inner.lock() {
+            Ok(t) => t.is_closed(),
+            Err(_) => true, // Well... what can we do?
```

Comment:
I like that we're not just unwrapping and potentially panicking on the lock acquisitions, but I'm not sure what here is safe? I believe the majority of the time we'd get an error on a lock is if the lock is poisoned, which means we might be in a bad state? is it ok to assume that if self.inner is poisoned that the connection is closed and we should move forward with trying to reconnect?

Reply:
The scenario where this will happen is when the reconnect callback executes functions on the sidecar instance (only location which calls back to caller from within the lock). And ... during the reconnect the original sidecar instance is always closed (otherwise the reconnect wouldn't happen in the first place), so actually `true` is the right value to report there.

Reply:
Comment updated 😛